### DISCUSSION TITLE
Scriptlets don't work in uBO

### DISCUSSION THREAD

**Discussion by dportvine** [uBlockOrigin/uAssets#23855] [Q&A] [answered]
https://github.com/AdguardTeam/AdguardFilters/issues/179220

I can't skip pre-roll with uBO.

<details>
<summary>Video</summary>

https://github.com/uBlockOrigin/uAssets/assets/111344219/09b2b4b9-1c5c-472d-99b6-5d0d34cb85f1
</details>

`https://suspilne.media/culture/schedule/tv/`

```
cosmonova-broadcast.tv#%#//scriptlet('trusted-replace-node-text', 'script', 'app.init', '"skip":0', '"skip":-1')
cosmonova-broadcast.tv#%#//scriptlet('trusted-click-element', '#preroll-skip-button', '', '2000')

**Comment by garry-ut99** [comment]
> `https://suspilne.media/culture/schedule/tv/`

After clicking play button, there is no video anymore, (or maybe geo-locked), only some message appears (same happens with uBlock disabled):

<details>
<summary>SPOILER</summary>

![gif](https://github.com/uBlockOrigin/uAssets/assets/72945564/6f19c0c4-1211-4e96-9210-2299dd64e6ed)

</details>

**Reply by dportvine** [reply]
What i see. Try Ukraine VPN.

<details>
<summary>Video</summary>

https://github.com/uBlockOrigin/uAssets/assets/111344219/c86e4b30-c91e-4eed-90b8-5828b878e69d
</details>

**Reply by garry-ut99** [reply]
If you provide a free proxy which works with `cosmonova-broadcast.tv`, then I can try, otherwise I'm not interested wasting my time in trying 10000 free proxies none of which are working with that site.

**Reply by dportvine** [reply]
Works with **Urban VPN proxy** & i used **Disable WebRTC**.

**Reply by garry-ut99** [reply]
Ok, works, I did tests and replied at the bottom.

**Comment by D4niloMR** [comment, marked as answer]
Try:

```
cosmonova-broadcast.tv##+js(rpnt, script, "skip":0, "skip":-1)
cosmonova-broadcast.tv##+js(trusted-click-element, #preroll-skip-button.enabled, , 2000)
```

and maybe `||analytics.cosmonova.net^`

**Reply by dportvine** [reply]
This works for me, also works with 100 milliseconds. Can you add that to Annoyance? This is self-promotion.

**Reply by Yuki2718** [reply]
> and maybe `||analytics.cosmonova.net^`

https://github.com/easylist/easylist/commit/0cdd73ee668ec20bdf6b386bb743a5e791ea8a42

**Reply by garry-ut99** [reply]
> dportvine : I can't skip pre-roll with uBO.

AG and uBO `trusted-replace-node-text` differ in number of required parameters: 
- AG requires 4 parameters [LINK 1](https://github.com/AdguardTeam/Scriptlets/blob/master/wiki/about-trusted-scriptlets.md#trusted-replace-node-text)
- uBO requires 3 parameters [LINK 2](https://github.com/gorhill/uBlock/wiki/Resources-Library#trusted-replace-node-textjs-)

the second parameter in AG (`'app.init'` in the current case) needs to be skipped/removed in uBO.  (unless your or anybody else's expectations are that uBO should remove it automatically, when enocuntering a filter written in AG syntax, then it could be considered uBO's scriptlet compatibility bug.)

Besides, it's a similiar case like in the previous question:  https://github.com/uBlockOrigin/uBlock-issues/issues/2881#issuecomment-2130745520

**Reply by gorhill** [reply]
> it could be considered uBO's scriptlet compatibility bug

uBO's `trusted-replace-node-text` came out first. That second parameter in AG's version correspond to uBO's vararg `condition, pattern`.

**Reply by garry-ut99** [reply]
> uBO's `trusted-replace-node-text` came out first.

Ok, if that means incompatibility is on AG's side, let it be so.
By the way, just providing some info regarding the dates:
- in uBO implemented `21.05.2023` https://github.com/gorhill/uBlock/commit/41876336db48292de06707adfa5e97dab74297d2
- in AG implemented `06.06.2023` https://github.com/AdguardTeam/Scriptlets/commit/e626cf29c3db5508beaba356721f481f70781345

>  That second parameter in AG's version correspond to uBO's vararg `condition, pattern`.

Yes, I did know about `condition, pattern`, somehow I just completely forgot about it, in such case I correct myself: 
`textMatch` (`'app.init'`) should not be removed/skipped when converting from AG to uBO, as it corresponds to `condition, pattern` in uBO, the proper conversion will be: 
`cosmonova-broadcast.tv##+js(rpnt, script, "skip":0, "skip":-1, condition, app.init)`

It seems uBO and AG had different approach regarding that parameter, as in uBO it's as _optional_, and in AG it's as _required_, it seems it's the different approach what is the reason of the incompatibility.

**Reply by garry-ut99** [reply]
> https://github.com/uBlockOrigin/uAssets/commit/02eebbdfb5dea8b35f8315bf7f91d71add81fd83

Sometimes works (most of time), but sometimes not (currrently tested in Firefox only):
- when does work (in-video ads removed) the both scriptlets are displayed in the logger:
  - the text scriptlet says text replaced
  - the click scriptlet says button clicked
- when doesn't work (and in-video ad appear): 
   - there is no `rpnt` displayed in the logger at all, like it didn't execute
   - the click scriptlet does appear but is waiting for a button and times out like the dom button targeted by it never existed, maybe the button didn't appear because `rpnt` failed

Besides, in AG doesn't work at all for me, like mentioned in: https://github.com/AdguardTeam/AdguardFilters/issues/179220#issuecomment-2134260671

**Comment by D4niloMR** [comment]
> Can you add that to Annoyance? This is self-promotion.

@dportvine is it always self-promotion? AdGuard added in the regional list, maybe we can add to uBlock filters.

**Reply by dportvine** [reply]
>is it always self-promotion?

Yes, this is one company. Such self-promotion is simply very annoying. You can also add it to uBlock filters

**Reply by D4niloMR** [reply]
https://github.com/uBlockOrigin/uAssets/commit/02eebbdfb5dea8b35f8315bf7f91d71add81fd83
